### PR TITLE
CreateDispatchers

### PR DESCRIPTION
**Pull request by MishkaSorochinskiy**

### CODE REVIEW

File: DeratControl.API/Dispatchers/CommandDispatcher.cs [hunk 1]
```diff
@@ -1,11 +1,48 @@
...
+using DeratControl.Application.Requests;
+using DeratControl.Domain.Entities;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
 
 namespace DeratControl.Application.Interfaces
 {
     public class CommandDispatcher
     {
+        private IHttpContextAccessor _context;
```
Comment (SviatoslavMykhailiv) [inline DeratControl.API/Dispatchers/CommandDispatcher.cs]: make it readonly

File: DeratControl.API/Dispatchers/CommandDispatcher.cs [hunk 2]
```diff
@@ -1,11 +1,48 @@
...
 
+        public CommandDispatcher(IHttpContextAccessor context)
+        {
+            this._context = context;
+        }
+
+        async Task<CommandResult>  Dispatch<TRequest>(TRequest command)where TRequest : IRequest
+        {
+            if (command == null)
```
Comment (SviatoslavMykhailiv) [inline DeratControl.API/Dispatchers/CommandDispatcher.cs:23]: you don't need else statement, "throw" will terminate the method . Make things simplier

File: DeratControl.API/Dispatchers/CommandDispatcher.cs [hunk 3]
```diff
@@ -1,11 +1,48 @@
...
+                throw new ArgumentNullException(nameof(command),
+                                                "Command can not be null.");
+            }
+
+            else if (!this._context.HttpContext.User.Identity.IsAuthenticated)
+            {
+                throw new UnauthorizedAccessException();
+            }
+
+            ICommandHandler<TRequest>handler= (ICommandHandler<TRequest>)this._context.HttpContext.RequestServices.
```
Comment (SviatoslavMykhailiv) [inline DeratControl.API/Dispatchers/CommandDispatcher.cs]: use 'var' where possible

File: DeratControl.API/Dispatchers/CommandDispatcher.cs [hunk 4]
```diff
@@ -1,11 +1,48 @@
...
+                GetService(typeof(ICommandHandler<TRequest>));
+
+            int userId =int.Parse(this._context.HttpContext.User.Identity.Name);
```
Comment (SviatoslavMykhailiv) [inline DeratControl.API/Dispatchers/CommandDispatcher.cs]: Name isn't an integer value, you'll encounter an error here

File: DeratControl.Application/Requests/CommandExecutionContext.cs [hunk 5]
```diff
@@ -1,10 +1,12 @@
 ﻿using System;
 using System.Collections.Generic;
 using System.Text;
+using DeratControl.Domain.Entities;
 
 namespace DeratControl.Application.Requests.Interfaces
 {
     public class CommandExecutionContext
     {
+        public User RequestedUser { get; set; }
```
Comment (SviatoslavMykhailiv) [inline DeratControl.Application/Requests/CommandExecutionContext.cs]: Once CommandExecutionContext is created, you cannot change RequestedUser

File: DeratControl.API/Startup.cs [hunk 6]
```diff
@@ -1,21 +1,34 @@
...
 
         // This method gets called by the runtime. Use this method to add services to the container.
         // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
         public void ConfigureServices(IServiceCollection services)
         {
-        }
+            services.AddDbContext<DeratContext>(options =>
+            options.UseSqlServer(_config.GetConnectionString("DefaultConnection")));
+
+            services.AddTransient(typeof(IHttpContextAccessor));
```
Comment (SviatoslavMykhailiv) [inline DeratControl.API/Startup.cs]: 1) Asp.net runtime will instantiate IHttpContextAccessor , you don't need to register it

File: DeratControl.API/Dispatchers/CommandDispatcher.cs [hunk 7]
```diff
@@ -1,11 +1,48 @@
...
     public class CommandDispatcher
     {
+        private IHttpContextAccessor _context;
 
+        public CommandDispatcher(IHttpContextAccessor context)
+        {
+            this._context = context;
+        }
+
+        async Task<CommandResult>  Dispatch<TRequest>(TRequest command)where TRequest : IRequest
```
Comment (SviatoslavMykhailiv) [inline DeratControl.API/Dispatchers/CommandDispatcher.cs]: public

Comment (SviatoslavMykhailiv) [reply DeratControl.API/Dispatchers/CommandDispatcher.cs]: fix formatting

File: DeratControl.API/Dispatchers/CommandDispatcher.cs [hunk 8]
```diff
@@ -1,11 +1,48 @@
...
+
+            ICommandHandler<TRequest>handler= (ICommandHandler<TRequest>)this._context.HttpContext.RequestServices.
+                GetService(typeof(ICommandHandler<TRequest>));
+
+            int userId =int.Parse(this._context.HttpContext.User.Identity.Name);
+
+            IRepository<User, int> userRepo =(IRepository<User, int>)this._context.HttpContext.RequestServices.
+                GetService(typeof(IRepository<User,int>));
+
+            User currentUser = userRepo.FindById(userId);
```
Comment (SviatoslavMykhailiv) [inline DeratControl.API/Dispatchers/CommandDispatcher.cs]: possible null value if a user hasn't been found

**Review by SviatoslavMykhailiv — CHANGES_REQUESTED** [review]

File: DeratControl.API/Dispatchers/CommandDispatcher.cs (hunk 1 above)
Comment (MishkaSorochinskiy) [reply DeratControl.API/Dispatchers/CommandDispatcher.cs]: Done) but i am not sure whether it is ok to parse Identity.Name in order to get our requested user)

File: DeratControl.API/Dispatchers/CommandDispatcher.cs (hunk 4 above)
Comment (MishkaSorochinskiy) [reply DeratControl.API/Dispatchers/CommandDispatcher.cs]: Done) But i am still not sure whether it works good

File: DeratControl.Application/Requests/CommandExecutionContext.cs [hunk 9]
```diff
@@ -7,6 +7,12 @@ namespace DeratControl.Application.Requests.Interfaces
 {
     public class CommandExecutionContext
     {
-        public User RequestedUser { get; set; }
+        private User requestedUser;
+        public User RequestedUser { get;}
```
Comment (SviatoslavMykhailiv) [inline DeratControl.Application/Requests/CommandExecutionContext.cs:10]: do you need a private field for that? Read only properties can be set in a constructor

**Review by SviatoslavMykhailiv — CHANGES_REQUESTED** [review]

**Review by SviatoslavMykhailiv — APPROVED** [review]
so far so good

**Pull request merged**